flink-tests/src/test/java/org/apache/flink/test/recovery/TaskManagerDisconnectOnShutdownITCase.java

```diff
@@ -0,0 +1,304 @@
+                                .createSessionComponentFactory(
+                                        new TestingStandaloneResourceManagerFactory(tracker));
+                    }
+                }) {
+            clusterEntrypoint.startCluster();
+
+            TestProcessBuilder taskManagerProcessBuilder =
+                    new TestProcessBuilder(TaskExecutorProcessEntryPoint.class.getName());
+            taskManagerProcessBuilder.addConfigAsMainClassArgs(config);
+
+            // start the TaskManager processes
+            taskManagerProcess = taskManagerProcessBuilder.start();
+
+            tracker.waitForTaskManagerConnected();
+
+            // shutdown TaskManager
+            taskManagerProcess.destroy();
+
+            tracker.waitForTaskManagerDisconnected();
+
+            assertEquals(
+                    "only one TaskManager should connect",
+                    1,
+                    tracker.getNumberOfConnectedTaskManager());
+        } catch (Exception e) {
+            e.printStackTrace();
+            printProcessLog(taskManagerProcess);
+            fail(e.getMessage());
+        } catch (Error e) {
+            e.printStackTrace();
+            printProcessLog(taskManagerProcess);
+            fail(e.getMessage());
+            throw e;
+        } finally {
+            if (taskManagerProcess != null) {
+                taskManagerProcess.destroy();
+                if (!taskManagerProcess.getProcess().waitFor(30, TimeUnit.SECONDS)) {
+                    log.error("TaskManager did not shutdown in time.");
+                    printProcessLog(taskManagerProcess);
+                    taskManagerProcess.getProcess().destroyForcibly();
+                }
+            }
+        }
+    }
+
+    protected static void printProcessLog(TestProcessBuilder.TestProcess process) {
+        if (process == null) {
+            System.out.println("-----------------------------------------");
+            System.out.println(" TaskManager WAS NOT STARTED.");
+            System.out.println("-----------------------------------------");
+        } else {
+            System.out.println("-----------------------------------------");
+            System.out.println(" BEGIN SPAWNED PROCESS LOG FOR TaskManager");
+            System.out.println("-----------------------------------------");
+            System.out.println(process.getErrorOutput().toString());
+            System.out.println("-----------------------------------------");
+            System.out.println("		END SPAWNED PROCESS LOG");
+            System.out.println("-----------------------------------------");
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+
+    /**
+     * The entry point for the TaskExecutor JVM. Simply configures and runs a TaskExecutor. Copy
+     * from AbstractTaskManagerProcessFailureRecoveryTest.
+     */
+    public static class TaskExecutorProcessEntryPoint {
```

Comment:
again, can we reuse the one from `AbstractTaskManagerProcessFailureRecoveryTest`?

Reply:
how about we refactor AbstractTaskManagerProcessFailureRecoveryTest into a standalone class?